src/main/kotlin/mate/academy/ArrayList.kt

```diff
@@ -1,44 +1,155 @@
 package mate.academy
 
+//import jdk.internal.util.ArraysSupport;
+import java.util.Arrays
+import java.util.NoSuchElementException
+
+
 class ArrayList<T> : List<T> {
+
+
+
+    private val defCapacity = 10
+    private val empty_arrayData = arrayOfNulls<Any>(0)
+
+
+    private var array: Array<T?> = arrayOfNulls<Any>(defCapacity) as Array<T?>
+    private var size = 0
+    private var cursor = 0
+
+
+    private fun errorMsg(index: Int): String {
+        return "Index: $index, Size: $size"
+    }
+
+
+    private fun grow(): Array<T?> {
+        val oldCapacity = array.size
+        val multiplicator = if (oldCapacity == size) 1.5 else 1.0
+        val newCapacity = (oldCapacity * multiplicator + 1).toInt()
+        array = if (oldCapacity > 0 || array !== empty_arrayData) {
+            return Arrays.copyOf(array, newCapacity)
+        } else {
+            return arrayOfNulls<Any>(defCapacity) as Array<T?>
+        }
+        return array
+    }
+
+    companion object {
+        fun listToArray(list: List<*>): Array<Any?>? {
+            if (!list.isEmpty()) {
+                return Array(list.size()) { index -> list[index] }
+            }
+            return null
+        }
+    }
+
     override fun add(value: T?) {
-        // implementation
+        val index = cursor
+        this.add(value, index)
+        cursor = index + 1
     }
 
+
     override fun add(value: T?, index: Int) {
-        // implementation
+        if (index > size || index < 0) {
+            throw ArrayListIndexOutOfBoundsException(errorMsg(index))
+        } else {
+            val s = size;
+            if (s == array.size) {
+                array = grow()
+            }
+            System.arraycopy(array, index, array, index + 1, size - index)
+            array[index] = value
+            size = s + 1
+        }
     }
 
+
     override fun addAll(list: List<T>) {
-        // implementation
+        if (!list.isEmpty()) {
+                val tempArray = listToArray(list) as Array<T?>
+            val newCapacity = size + tempArray.size
+            array = Arrays.copyOf(array, newCapacity)
+            System.arraycopy(tempArray, 0, array, size, tempArray.size)
+            size = newCapacity
+        }
```

Comment:
Looks like you are using too many copy and converting data methods
It can be more effective even with only one for cycle, which will rewerite data to new array with needed size